### ISSUE TITLE
Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope

### ISSUE THREAD

**Issue by RoopanV** [module-federation/module-federation-examples#1142]
Hi,
     First thanks a lot for giving us module federation. It is working really well for me. I was inspired by the dynamic remote example and I tried to implement that because i have 22 mfe and loading 22 remoteEntry.js file at once increases the initial render time. I followed the same approach as mentioned in the example but facing the above issue. My scenario also bit weird such that one mfe is library kind of thing. So all other 21 mfe use that library mfe.

Config for shell:

```
new ModuleFederationPlugin({
      name: 'shell',
      remotes: {
        app_library: app_library@app_library/remoteEntry.js,
      },
     .......
  })
```

Config for remaning MFEs:

```
  new ModuleFederationPlugin({
      name: <respective name>,
      filename: 'remoteEntry.js',
      library: { type: 'var', name: <respective name>},
      exposes: {
        ...exposedPaths,
      },
      remotes: {
        app_library: 'app_library',
      },
      },
    })
```

Config for app_library:

```
 new ModuleFederationPlugin({
      name: app_library,
      filename: 'remoteEntry.js',
      library: { type: 'var', name: 'app_library' },
      exposes: {
        ...exposedPaths,
      },
    })
```

Low level api's used:

```
 await __webpack_init_sharing__('default')
  const container = (window as any)[remoteName] as Container
  await container.init(__webpack_share_scopes__.default)
  const factory = await container.get(exposedModule)
  const Module = factory()
  return Module as T
```
**Issue Facing:**

On loading Shell alone things (which loads app_library mfe too) works fine. But when I try to load other mfe through lazy routing it is throwing below error because app_library is already loaded. And stops there without navigating further.

`sharing:17 Initialization of sharing external failed: Error: Container initialization failed as it has already been initialized with a different share scope`

**Expectation:**

Instead of trying to initialize again it should use already initialized with or without warning and proceed. 

**Note:**

Not sure it is a bug  or feature request or question for stack overflow. Since I don't find any example for my scenario posting here. 

Thanks in advance

**Comment by ScriptedAlchemy** [comment]
you cant initialize a container more than once, put it in a try catch block or if using the low level API add something like window.remoteContainer.__initialized = true that way you can check if its already been initialized or not

**Comment by RoopanV** [comment]
Hi @ScriptedAlchemy, I agree that we can't initialise a container more than once. Here I am only intialising a mfe (app_mfe1) which internally initialising the other one (app_library). And that code is generated by webpack in remoteEntry.js
Can we pass something to override that one?

**Comment by ScriptedAlchemy** [comment]
Only a try catch block around the initialize function. It cannot be overridden and shouldn't be

**Comment by dengbupapapa** [comment]
> Hi, First thanks a lot for giving us module federation. It is working really well for me. I was inspired by the dynamic remote example and I tried to implement that because i have 22 mfe and loading 22 remoteEntry.js file at once increases the initial render time. I followed the same approach as mentioned in the example but facing the above issue. My scenario also bit weird such that one mfe is library kind of thing. So all other 21 mfe use that library mfe.
> 
> Config for shell:
> 
> ```
> new ModuleFederationPlugin({
>       name: 'shell',
>       remotes: {
>         app_library: app_library@app_library/remoteEntry.js,
>       },
>      .......
>   })
> ```
> 
> Config for remaning MFEs:
> 
> ```
>   new ModuleFederationPlugin({
>       name: <respective name>,
>       filename: 'remoteEntry.js',
>       library: { type: 'var', name: <respective name>},
>       exposes: {
>         ...exposedPaths,
>       },
>       remotes: {
>         app_library: 'app_library',
>       },
>       },
>     })
> ```
> 
> Config for app_library:
> 
> ```
>  new ModuleFederationPlugin({
>       name: app_library,
>       filename: 'remoteEntry.js',
>       library: { type: 'var', name: 'app_library' },
>       exposes: {
>         ...exposedPaths,
>       },
>     })
> ```
> 
> Low level api's used:
> 
> ```
>  await __webpack_init_sharing__('default')
>   const container = (window as any)[remoteName] as Container
>   await container.init(__webpack_share_scopes__.default)
>   const factory = await container.get(exposedModule)
>   const Module = factory()
>   return Module as T
> ```
> 
> **Issue Facing:**
> 
> On loading Shell alone things (which loads app_library mfe too) works fine. But when I try to load other mfe through lazy routing it is throwing below error because app_library is already loaded. And stops there without navigating further.
> 
> `sharing:17 Initialization of sharing external failed: Error: Container initialization failed as it has already been initialized with a different share scope`
> 
> **Expectation:**
> 
> Instead of trying to initialize again it should use already initialized with or without warning and proceed.
> 
> **Note:**
> 
> Not sure it is a bug or feature request or question for stack overflow. Since I don't find any example for my scenario posting here.
> 
> Thanks in advance

Excuse me, do you have a solution？

**Comment by ScriptedAlchemy** [comment]
Try catch the init or use webpack_require.l for chunk loading

**Comment by ScriptedAlchemy** [comment]
```js

try {
 await container.init(__webpack_share_scopes__.default)
} catch(e) {
// already was initialized
}
```

**Comment by marcosfilho-hotmart** [comment]
@ScriptedAlchemy shouldn't this be a default behavior, or at least configured via plugin flag instead of an internal api configuration? That is, if a container was already initialized on an old scope, console warn that an old scope was already present and utilize this existing scope instead of creating a new one. Do you think there would be any side effects with this behavior?

Moreover, this really complicates the use of module federation containers inside library exports or module federation containers since reloading the container on a different route it's a fairly common scenario and most developers like myself don't know yet how to use the internal api and where to adjust this container.init call.


To expand more on my scenario:

I have a container project App1 that is not a module federation container, and that consumes a secondary App2 project via UMD export on a Shadow Dom. This App2 provides three routes to be consumed, and in one of those it consumes an App3 project via module federation, that is, App1 (host) -> App2 (umd build) -> App3 (module federation). When the host app re-loads App2 on a different route, if this route imports App3, then it will throw the aforementioned error. This feels like a fairly common gotcha, to reload the App2 if a route changes on App1, and while I do realize reloading the App2 is not ideal, that is the reality on many projects I work with and changing would require a huge effort.

**Comment by ScriptedAlchemy** [comment]
No, since this is considered a user error that you are trying to initialize share scopes more than once into a container thats already been initialized - so you shouldnt be calling the init method multiple times. 

If you do, webpack is going to perform a hard crash as this could lead to unforseen side effects, so we throw an error.

Imagine a case where you try to pass two completely different share scopes, and there was no error about it. Things would work/break without any explanation. 

So the issue here is load component is too primitive. 


```js
        var __webpack_error__ = new Error();
        module.exports = new Promise(function (resolve, reject) {
          if (typeof home !== 'undefined') return resolve();
          __webpack_require__.l(
            'http://localhost:3001/_next/static/chunks/remoteEntry.js',
            async function (event) {
              await home.init(__webpack_share_scopes.default)
              if (typeof home !== 'undefined') return resolve();
              var errorType = event && (event.type === 'load' ? 'missing' : event.type);
              var realSrc = event && event.target && event.target.src;
              __webpack_error__.message =
                'Loading script failed.\n(' + errorType + ': ' + realSrc + ')';
              __webpack_error__.name = 'ScriptExternalLoadError';
              __webpack_error__.type = errorType;
              __webpack_error__.request = realSrc;
              reject(__webpack_error__);
            },
            'home',
          );
        }).then(function () {
          return home;
        });

        ```
        
        would be a better way to do it, then webpack is tracking if chunks have already been loaded or not

**Comment by AndreiSoroka** [comment]
Hi! We have MF applications: HOST, A, B, C

Where dependencies:
`HOST <- A`
`HOST <- B`
`A <- C`
`B <- C`

<img width="412" alt="image" src="https://user-images.githubusercontent.com/6079139/182334391-2f9ed7be-95b2-471b-954d-aeab1dc0d1ef.png">

So, when we load `HOST <- A <- C`, all is ok. After that when we change router (for example) and load  `HOST <- B <- C` then we catch error
```
Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope
```
Because module `C` was initialized before. But I don't understand, why do we think that these are different scopes? Do Module Federation support this case from a plugin?

<img width="1601" alt="image" src="https://user-images.githubusercontent.com/6079139/182335658-0d115986-8ea9-4a37-a664-33a510ab0d4f.png">

So, please, can you help find a solution to how I can resolve it?

P.s. my vision: it is should support by the MF plugin. We have one URL for bundle module `C`, and we have one scope (maybe need to deep check, but not equal link to object?).

**Comment by marcosfilho-hotmart** [comment]
**Edit:** 
We managed to mostly fix this issue for our enterprise context by creating a federated app loader, I will be commenting the solution here later if others are interested, but I still have an issue. On a scenario where a host App1 loads a UMD library App2, which in turn loads App3 as MF container, if App2 is destroyed and then reconstructed, then App3 no longer loads styles. This is probably an intended behavior as the module has already been loaded previously (and is now available at window['app-name']), but we need a way to force the reload of styles again since they were previously inserted on a Shadow Dom (via style-loader) but that was destroyed when the app was initially destroyed. This behavior prevents the usage of Module Federation inside Shadow Dom containers.



**Original post:**
I see, unfortunately I have a complex scenario where the app could be rendered as a UMD library or a module federation container, and this complicates the build process by quite a bit. I was trying to avoid loading the module federation via URL directly and load it as the plugin's alias, but it doesn't really work :/

I believe the module's entrypoint is loaded by default and the function init is called by webpack's ensure function if it identifies the module might be loaded somewhere as an async import, because we define a map of app names and module federation definition names to import it using the alias provided by the ModuleFederation remotes property and the file is clearly downloaded on the Network window even if there was no call for it to load.

Scenarios:

**Host App1 loads -> UMD App2 loads -> Module Federation App3
Host App1 loads -> Module Federation App2 -> Module Federation App3**

I'm not sure i can clearly explain my context, but it's something like this:

Webpack config:

```javascript
plugins: [
    new ModuleFederationPlugin({
      name: 'app_1,
      filename: 'remoteEntry.js',
      exposes: {
        './Main': join(PATHS.STARTUP, 'module-federation/index.tsx'),
      },
      remotes: {
        App2: 'app_2@http://local.app-route:5001/remoteEntry.js'
      },
      shared: {
        react: {
          requiredVersion: pkjson.dependencies.react,
        },
        'react-dom': {
          requiredVersion: pkjson.dependencies['react-dom'],
        },
        'react-router-dom': {
          requiredVersion: pkjson.dependencies['react-router-dom'],
        }
      }
    }
```
Code that loads the module federation

```javascript
  // this triggers a preload by webpack to ensure the module exists, which automatically calls __init__ function of the module federation container and registers it as `window[appName]`. This happens even if getRenderFedereatedApp executes
  const appList = {
    'app_2': () => import('App2/Main').then(({ default: render }) => render)
  }. 

  const getRenderFederatedApp = async (appName: string, appModule: string): Promise<FederatedAppRender> => {
    try {
      // If module has already been loaded, then just get the factory reference
      const factory = await window[appName]?.get(appModule)
      
      if (factory) {
        const { default: render } = factory()

        return render
      }
     
     // load app if it hasn't been loaded yet, and return the render function. This should be modified to the script loading version
     return await appList[appName]?.()

      throw new Error(`Missing federated module render function for app: ${appName}`)
    } catch (error) {
      console.error(`App ${appName} was not imported correctly, check the module import key or the module entrypoint file. \n\n${error} `)

      Promise.reject(error)
    }

    return Promise.reject(new Error())
  }

  return { getRenderFederatedApp, getAppRoot, setAppRoot }
}
```

It feels like our problem is similar to Andrey's, in that this happens mostly when we change routes on the Host parent and that triggers a re-instantiation of the App3, either because another App4 utilizes the same container or the same App2 is re-instantiated (routing happens on host and that completely reload the App2)

**Comment by jegor-cloudmore** [comment]
Have similar problem, where dependencies:
HOST(MAIN) <- HOST/REMOTE(A) <- REMOTE(C)
HOST(MAIN) <- HOST/REMOTE(B) <- REMOTE(C)

HOST/REMOTE uses webpack config.
REMOTE uses webpack config.
HOST uses vite config.

When component C called second time from different remote, then error is thrown.

`Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`

**Comment by AndreiSoroka** [comment]
Hi!
@ScriptedAlchemy do you have any idea?

**Comment by AndreiSoroka** [comment]
P.s. and, can we reopen this issue? 
FYI @ScriptedAlchemy

**Comment by marcosfilho-hotmart** [comment]
@AndreiSoroka @ScriptedAlchemy Here are some reference files for people that still require help with this issue. 

**This setup enables App export as a UMD library and a Module Federated container. Said App is able to handle the following flows (consider UMD modules as Shadow Dom containers):**

- Module Federated App1 imports -> Module Federated App2
- Module Federated App1 imports  -> Module Federated App2 imports -> Module Federated App 3
- Module Federated App1 imports  -> Module Federated App2 imports -> UMD Module App 3
- Module Federated App1 imports -> UMD Module App2
- Module Federated App1 imports -> UMD Module App2 imports -> UMD Module App3
- UMD Module App1 imports -> Module Federated App2
- UMD Module App1 imports -> UMD Module App2 imports -> UMD Module App3

**This setup DOES NOT enable the following scenarios**
- UMD Module App1 imports -> Module Federated App2 imports -> anything

This issue happens because Module Federated styles are only imported once, and upon Shadow Dom destruction the styles located inside the Shadow Dom are destructed and never again imported as the module has already been loaded and processed.

- Webpack production config: https://gist.github.com/marcosmapf/a1c0b83a3f177462e8bf604ceba38323
- Webpack StyleTagTransform custom script: https://gist.github.com/marcosmapf/904986dde911148e39a17a59dc3dd132
- App index.js: https://gist.github.com/marcosmapf/fc9bd8d1408c14a637fa0e280cff2732
- Module-Federated App bootstrap.tsx: https://gist.github.com/marcosmapf/49fbe90d2061183cc18789225ecd9dc4
- Module Federation app util handler: https://gist.github.com/marcosmapf/4ab9c30dfd78656bb4d6c4ca33e239e3
- UMD App bootstrap.tsx: https://gist.github.com/marcosmapf/abf5a396f7404459418cf008cf160149

**Comment by AndreiSoroka** [comment]
It's like a jok )
I tryed to reproduce this bug, but oll work is fine (about concept A->B1,B2->C)

You can check here: https://github.com/AndreiSoroka/mf-reproduce-collision-names

But i found other bug with name collision, so new issue: https://github.com/module-federation/module-federation-examples/issues/2201

**Comment by ScriptedAlchemy** [comment]
Are two remotes called the same thing? Is it one remote overriding the global of another? 

Have you tried setting uniqueName on webpack config to same name as MF name

**Comment by AndreiSoroka** [comment]
> Are two remotes called the same thing?

yes, the component name:`HomeView`

- Host component  `HomeView.vue`: https://github.com/AndreiSoroka/mf-reproduce-collision-names/blob/master/service-a/src/views/HomeView.vue
- Connected remotely component `b3/HomeView.vue': ` https://github.com/AndreiSoroka/mf-reproduce-collision-names/blob/master/service-a/src/views/B3View.vue

> Is it one remote overriding the global of another?

No

> Have you tried setting uniqueName on webpack config to same name as MF name

What do you mean? Unique component name? ModuleFederationPlugin > name ? ModuleFederationPlugin > exposes > component Name?

**Comment by xuebigshuai** [comment]
P.s. and, can we reopen this issue? I also have this problem.
How did it work out in the end?

**Comment by xuebigshuai** [comment]
it's not good in vite project 。 But it's good in vue-cli(webpack) project

**Comment by xuebigshuai** [comment]
> Have similar problem, where dependencies: HOST(MAIN) <- HOST/REMOTE(A) <- REMOTE(C) HOST(MAIN) <- HOST/REMOTE(B) <- REMOTE(C)
> 
> HOST/REMOTE uses webpack config. REMOTE uses webpack config. HOST uses vite config.
> 
> When component C called second time from different remote, then error is thrown.
> 
> `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`

Solved it？ in vite host

**Comment by erkanisuf** [comment]
I am also having same issue:
`Uncaught (in promise) Error: Container initialization failed as it has already been initialized with a different share scope`

I do use: /remoteEntry.js?v=[Date.now()]

I got Host app (A).
Component apps(B)
Component apps(C)
Component apps(D)

Component apps B and D share components between each other and also the host app takes components from ABCD. Not sure if this happens because of caching or what. B C D are deployed to S3 with cloud front cache to set none. Meanwhile Host App is in heroku with express JS cache set to none.

**Comment by ScriptedAlchemy** [comment]
When using federation remote containers, it is important to ensure that they are initialized only once. Otherwise, an error will be thrown and the container will not work properly. One way to avoid this problem is to wrap the container.init() function call in a try catch block. This way, if the container has already been initialized, the error will be caught and ignored. Another way is to add a marker property to the container after initialization, and check for its existence before calling init() again. Either method can help you safely use federation remote containers in your code.

**Comment by erkanisuf** [comment]
Is it possible to have some kind of basic example as i feel really lost in this 😿 
The way i fetch remotes is in the webpack with the plugin:

```
	new ModuleFederationPlugin(
			{
				name: 'NAMEHERE',
				filename:
					'remoteEntry.js',
				remotes: {
					B: 'url'
					C: 'url'.
					...etc..
				},
				shared: {
					...deps}
```

I also checked the function of getOrLoadRemote  (I assume it takes care to check if remote is already fetched) in [https://gist.github.com/ScriptedAlchemy/3a24008ef60adc47fad1af7d3299a063](url)

But i don't understand where to trigger this function exactly.
Ive tried to use it in the `remotes:{} `object but seems that its not the place for it.
Application is in React.

**Comment by marcosfilho-hotmart** [comment]
> Another way is to add a marker property to the container after initialization, and check for its existence before calling init() again. Either method can help you safely use federation remote containers in your code.

Is it safe to use the window reference for the module, such as `window.federeated_app_name`, as the marker for if it has already been loaded? We also add a "root" property to `window.federated_app_name` after loading to indicate where the module is currently used, which enables us to properly scope CSS and search for children if the module is inside a shadow dom. I have posted an example of this solution on a previous comment

**Comment by AburronKing** [comment]
> 使用联合远程容器时，确保它们仅初始化一次非常重要。否则，将会抛出错误并且容器将无法正常工作。避免此问题的一种方法是将 container.init() 函数调用包装在 try catch 块中。这样，如果容器已经初始化，错误将被捕获并忽略。另一种方法是在初始化后向容器添加一个标记属性，并在再次调用 init() 之前检查其是否存在。任一方法都可以帮助您在代码中安全地使用联合远程容器。

I don't know where or how to configure【container.init()  or   try catch】，can you show in vue-cli?

**Comment by AburronKing** [comment]
My problem is：in micro app,2 Host Use a remote at the same time, 
report error：

`Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`

**Comment by marcosfilho-hotmart** [comment]
@AburronKing I believe this happens because you are trying to initialize the module twice using different share scopes. You might need to verify if the module has been loaded before trying to initialize it again, and if it has been loaded, then utilize the same factory function. That can be done by checking the window reference for the module, and verifying if it has a factory function. I created a gist about this:

https://gist.github.com/marcosmapf/4ab9c30dfd78656bb4d6c4ca33e239e3#file-federatedapp-ts-L33

**Comment by AburronKing** [comment]
> @AburronKing我相信发生这种情况是因为您尝试使用不同的共享范围两次初始化模块。您可能需要在尝试再次初始化模块之前验证模块是否已加载，如果已加载，则使用相同的工厂函数。这可以通过检查模块的窗口引用并验证它是否具有工厂功能来完成。我为此创建了一个要点：
> 
> https://gist.github.com/marcosmapf/4ab9c30dfd78656bb4d6c4ca33e239e3#file-federatedapp-ts-L33

Thank you very much

**Comment by AburronKing** [comment]
> My problem is：in micro app,2 Host Use a remote at the same time, report error：
> 
> `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`

When I upgraded the webpack version to 5.88.2 for all applications, the issue was resolved.

**Comment by xuebigshuai** [comment]
> > My problem is：in micro app,2 Host Use a remote at the same time, report error：
> > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> 
> When I upgraded the webpack version to 5.88.2 for all applications, the issue was resolved.

兄den, 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。

**Comment by xuebigshuai** [comment]
> > @AburronKing我相信发生这种情况是因为您尝试使用不同的共享范围两次初始化模块。您可能需要在尝试再次初始化模块之前验证模块是否已加载，如果已加载，则使用相同的工厂函数。这可以通过检查模块的窗口引用并验证它是否具有工厂功能来完成。我为此创建了一个要点：
> > https://gist.github.com/marcosmapf/4ab9c30dfd78656bb4d6c4ca33e239e3#file-federatedapp-ts-L33
> 
> Thank you very much

这个怎么个用法？

**Comment by AburronKing** [comment]
> > > My problem is：in micro app,2 Host Use a remote at the same time, report error：
> > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > 
> > 
> > When I upgraded the webpack version to 5.88.2 for all applications, the issue was resolved.
> 
> 兄den, 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。

展示一下你的核主要代码

**Comment by ScriptedAlchemy** [comment]
You need to try catch the init. It can only fire once and dynamic remotes tear share scope object as they initialize and seal right away. In webpack init is passed a second argument which is array used to track all remotes have provided share scoped. Then they all seal. 

We are close to releasing a major upgrade to federation. In it we will have better runtime APIs where these issues will likely no longer exist and you'll have helpers available from my packages. https://github.com/module-federation/universe/pull/1623

**Comment by xuebigshuai** [comment]
> > > > My problem is：in micro app,2 Host Use a remote at the same time, report error：
> > > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > > 
> > > 
> > > When I upgraded the webpack version to 5.88.2 for all applications, the issue was resolved.
> > 
> > 
> > 兄den, 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。
> 
> 展示一下你的核主要代码

我重写了下 remote 引用时的init和get ，当相同remote再次被引用时，不再走初始化逻辑，只初始化一次就不会出现共享容器报错的这个问题

**Comment by xuebigshuai** [comment]
> > > > 我的问题是：在微应用中，2台主机同时使用遥控器，报错：
> > > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > > 
> > > 
> > > 当我将所有应用程序的 webpack 版本升级到 5.88.2 时，问题得到了解决。
> > 
> > 
> > 兄den， 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。
> 
> 展示一下你的核主要代码

![image](https://github.com/module-federation/module-federation-examples/assets/22486789/2ea8197d-5bf8-4e73-9d0e-77a216ece8fe)
通过重写remote的使用方式

**Comment by aplanchamp** [comment]
Hello, I am currently bumping into this issue from times to times and I really can't figure out why.
I have created a repository to reproduce : [https://github.com/aplanchamp/poc-mfe](https://github.com/aplanchamp/poc-mfe)

**Context :** 

A next 13 shell application integrating 3 react components as microfrontends. 

- the shell application and the react MFEs are all in version 18.2.0
- the shell application and the react MFEs share a common package "context-manager" which is a singleton (setup in the module federation configuration)

When running the shell application I get this inconsistent issue during navigation. When that occurs, the application reloads automatically and everything works fine, but still, I would like to prevent this issue. 

**Versions :** 

"next": "13.5.6",
"react": "18.2.0",
"react-dom": "18.2.0"
"@module-federation/nextjs-mf": "^8.0.1-4" (needed to use next/router package)

**Getting started :**  

Steps are described in the readme file to run MFEs on client side only.

_Also, if you have time, there is also the steps to launch everything on server side in order to use "dynamic" rather than "lazy" to load the components. This actually works well and prevent having to use effect to run on client only (effect that is needed when using react.lazy otherwise you get a mismatch error, error that is also present on your 'next-react' example when you refresh the page by the way). For what reasons shouldn't we use dynamic (cause you specify to use lazy in readme) ? Also, would it be ok to handle server side rendering of the react component using dynamic and { ssr: true } (meaning handling SSR even if MFEs are not next but react) ? Right now I am getting issues with this strategy IF I use hooks on MFEs.  
Sorry, I realize that this last paragraph asks a lot of questions, should I create a new issue for this matter ?_

**Comment by UJPrasad** [comment]
@aplanchamp have you got a fix for this issue?

**Comment by wenjie79** [comment]
> > > > > 我的问题是：在微应用中，2台主机同时使用遥控器，报错：
> > > > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > > > 
> > > > 
> > > > 当我将所有应用程序的 webpack 版本升级到 5.88.2 时，问题得到了解决。
> > > 
> > > 
> > > 兄den， 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。
> > 
> > 
> > 展示一下你的核主要代码
> 
> ![image](https://private-user-images.githubusercontent.com/22486789/284066949-2ea8197d-5bf8-4e73-9d0e-77a216ece8fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5MTgyMzAsIm5iZiI6MTcyNDkxNzkzMCwicGF0aCI6Ii8yMjQ4Njc4OS8yODQwNjY5NDktMmVhODE5N2QtNWJmOC00ZTczLTlkMGUtNzdhMjE2ZWNlOGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDA3NTIxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4ZmNkOWE2NzA1YmEzN2JhNWViZDc0YzA5Njk0ZTVjMzEzM2Y0MTc2ODc2ZGM4MjFjOWNlZTM5OTkxMGQ3N2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AyHs1Nc8tfOV1Pp_ehaV3ukqViLMWpWJ5ipowgL2B60) 通过重写remote的使用方式

图片失效了，能否重新发一下谢谢

**Comment by wenjie79** [comment]
> > > > > 我的问题是：在微 app 中，2 Host 同时使用远程，报错：
> > > > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > > > 
> > > > 
> > > > 当我将所有应用程序的 webpack 版本升级到 5.88.2 时，问题已得到解决。
> > > 
> > > 
> > > 兄den， 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。
> > 
> > 
> > 展示一下你的核主要代码
> 
> 我重写了下 remote 引用时的init和get ，当相同remote再次被引用时，不再走初始化逻辑，只初始化一次就不会出现共享容器报错的这个问题

能否贴出你的代码?谢谢

**Comment by AburronKing** [comment]
> > My problem is：in micro app,2 Host Use a remote at the same time, report error：
> > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> 
> When I upgraded the webpack version to 5.88.2 for all applications, the issue was resolved.

main-app code:
![image](https://github.com/user-attachments/assets/b23c0a11-354a-4e43-b26c-ac627e580e26)
child-app code:
![image](https://github.com/user-attachments/assets/32286f23-ecd7-4e39-b939-a5eb73d80de1)

No more settings，just that

**Comment by huchaolin** [comment]
> > > > > 我的问题是：在微应用中，2台主机同时使用遥控器，报错：
> > > > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > > > 
> > > > 
> > > > 当我将所有应用程序的 webpack 版本升级到 5.88.2 时，问题得到了解决。
> > > 
> > > 
> > > 兄den， 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。
> > 
> > 
> > 展示一下你的核主要代码
> 
> ![image](https://private-user-images.githubusercontent.com/22486789/284066949-2ea8197d-5bf8-4e73-9d0e-77a216ece8fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNjQzNzgsIm5iZiI6MTcyNzI2NDA3OCwicGF0aCI6Ii8yMjQ4Njc4OS8yODQwNjY5NDktMmVhODE5N2QtNWJmOC00ZTczLTlkMGUtNzdhMjE2ZWNlOGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI1VDExMzQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQzNmIwODc3NjZiNDI4NGIxZWYzNzQyYmRhNTYyNDRhZDM2MWVjOTJhM2UzNDZmOGNhMWYwYzQ3ZTI5NmIyYzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.46sd1YHOslbfVhCvCfhERbNkFZlFlxhy7YHO32UPtRk) 通过重写remote的使用方式

代码地址 参考 https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes

**Comment by ScriptedAlchemy** [comment]
Use module federation v2.
We have built substantial infrastructure around federation now. 

https://module-federation.io/

**Comment by ScriptedAlchemy** [comment]
Or. Check my runtime plugins folder in this repo for user plugins. Promise new promise isn't a good solution and you can do better things via the new runtime plugin system we built. 

Most problems mentioned in this git issue no longer exist in v2.

**Comment by aplanchamp** [comment]
> @aplanchamp have you got a fix for this issue?

Hey @UJPrasad sorry for the late reply, no I did not find a fix with the versions mentioned in the issue, but we are using the version "8.1.7" of @module-federation/nextjs-mf and we don't experience the issue anymore.

**Comment by krulun** [comment]
> > > My problem is：in micro app,2 Host Use a remote at the same time, report error：
> > > `Dynamic remote throwing Container initialization failed as it has already been initialized with a different share scope`
> > 
> > 
> > When I upgraded the webpack version to 5.88.2 for all applications, the issue was resolved.
> 
> 兄den, 你这个问题升级了webpack到5.88.2 就解决了？ 我在qiankun 中使用模块联邦嵌套，也出现了这个问题。

最后你解决了 吗，怎么解决的
